File: pdoo/ruby/weapon.rb
#encoding:utf-8

require "./lib/WeaponToUI.rb"

module Deepspace
  # Armas de la estación
  class Weapon
    attr_reader :name, :type, :uses

    def initialize(weapon_name,weapon_type,weapon_uses)
      @name = weapon_name
      @type = weapon_type
      @uses = weapon_uses
    end

    def self.newCopy(weapon)
      new(weapon.name,weapon.type,weapon.uses)
    end
  
    def power
      @type.power
    end

    def useIt
      if @uses > 0
        @uses = @uses - 1
        return power
      else
        return 1.0
      end
    end

    def getUIversion
      WeaponToUI.new(self)
    end

    def to_s
      getUIversion.to_s
    end
  end
end

File: pdoo/java/Deepspace/Weapon.java
/**
 * @author Manuel Gachs Balleger
 */
package Deepspace;

/**
 * @brief Armas de las que dispone una estación
 */
class Weapon implements Copyable <Weapon>, CombatElement {
    private String name;
    private WeaponType type;
    private int uses;
    
    Weapon(String weaponName,WeaponType weaponType,int weaponUses) {
        name = weaponName;
        type = weaponType;
        uses = weaponUses;
    }
    Weapon(Weapon w) {
        name = w.name;
        type = w.type;
        uses = w.uses;
    }
    public WeaponType getType() {return type;}
    @Override
    public int getUses() {return uses;}
    public float power() {return type.getPower();}
    @Override
    public float useIt() {
        if (uses>0) {
            uses = uses - 1;
            return this.power();
        } else
            return 1.0f;
    }
    
    WeaponToUI getUIversion() { return new WeaponToUI(this); }
    
    @Override
    public String toString() {
        String aux = "Nombre: " + name;
        aux += ", Tipo: " + type;
        aux += ", Usos: " + uses;
        return aux;
    }

    @Override
    public Weapon copy() {
        return new Weapon(this);
    }
}